### PR TITLE
[Feat][Router] Add context signal for token count-based routing

### LINKED ISSUE #1106: [v0.2-Athena]: Add Context Signal for Token Count-Based Routing
## Description

Add a new **Context Signal** type to the signal-driven routing system that classifies requests based on token count. This signal enables intelligent routing decisions based on the size of the input context, allowing the system to route short requests to faster models and long-context requests to models with larger context windows.

## Context

Currently, the system supports 7 signal types (keyword, embedding, domain, fact_check, user_feedback, preference, language) for routing decisions. However, there's no built-in way to route based on the size of the request (token count), which is a critical factor for:

1. **Cost optimization**: Route short requests to cheaper, faster models
2. **Performance optimization**: Avoid using large-context models for small requests
3. **Context window management**: Ensure long requests go to models that can handle them
4. **Resource allocation**: Better distribute workload based on request complexity

### Current Signal System

The signal system allows combining multiple signals with AND/OR logic to make routing decisions:

```yaml
decisions:
  - name: "code_route"
    rules:
      operator: "AND"
      conditions:
        - type: "keyword"
          name: "code_keywords"
        - type: "domain"
          name: "computer_science"
```

### Proposed Context Signal

Add a new `context` signal type that classifies based on token count ranges:

```yaml
signals:
  context_rules:
    - name: "low_token_count"
      min_tokens: "0"
      max_tokens: "1K"
      description: "Short requests with less than 1K tokens"
    
    - name: "medium_token_count"
      min_tokens: "1K"
      max_tokens: "4K"
      description: "Medium requests between 1K-4K tokens"
    
    - name: "high_token_count"
      min_tokens: "4K"
      max_tokens: "128K"
      description: "Long requests with more than 4K tokens"

decisions:
  - name: "long_context_handler"
    rules:
      operator: "AND"
      conditions:
        - type: "context"
          name: "high_token_count"
        - type: "domain"
          name: "code"
    model_refs:
      - model: "claude-3-opus"  # Large context window model
```

### Key Design Decisions

1. **Use absolute token counts** (not percentages) - avoids dependency on target model's max context
2. **Support K/M suffixes** for better readability: `"1K"` = 1000, `"1M"` = 1000000
3. **Support decimal values**: `"1.5K"` = 1500 tokens
4. **Calculate before routing** - token counting happens during signal evaluation phase
5. **Use tokenizer** - leverage existing tokenizer infrastructure for accurate counting

### 1. Core Data Structures

- [ ] Add `SignalTypeContext = "context"` constant to `src/semantic-router/pkg/config/config.go`
- [ ] Create `TokenCount` custom type with `UnmarshalYAML` support for K/M suffixes
- [ ] Add `ContextRule` struct with `MinTokens` and `MaxTokens` fields
- [ ] Add `ContextRules []ContextRule` to `Signals` struct

### 2. Token Counting

- [ ] Create `src/semantic-router/pkg/classification/context_classifier.go`
- [ ] Implement `ContextClassifier` with `Classify(text string) ([]string, int, error)` method
- [ ] Add `TokenCounter` interface for pluggable tokenizer implementations
- [ ] Integrate with existing tokenizer (Candle UnifiedTokenizer or tiktoken)

### 3. Signal Integration

- [ ] Add `MatchedContextRules []string` to `SignalResults` struct
- [ ] Add `TokenCount int` field to `SignalResults` for observability
- [ ] Update `EvaluateAllSignals()` to evaluate context rules
- [ ] Add `ContextRules []string` to `SignalMatches` in decision engine

### 4. Decision Engine

- [ ] Add `case "context"` to signal matching logic in `evaluateRuleCombinationWithSignals()`
- [ ] Support context conditions in decision rules

### 5. Observability

- [ ] Add Prometheus metric `llm_context_token_count` histogram
- [ ] Add `RecordContextTokenCount(model, tokenCount, contextLevel)` function
- [ ] Expose token count in response headers: `X-VSR-Context-Token-Count`
- [ ] Log context classification results

### 6. API Types (Kubernetes CRD)

- [ ] Add `ContextRules []ContextRule` to `Signals` struct in `types_route.go`
- [ ] Add validation annotations for token count ranges
- [ ] Update CRD generation

### 7. Configuration & Documentation

- [ ] Update `config.yaml` with context signal examples
- [ ] Add context signal documentation to website
- [ ] Update signal-driven-decisions.md to include context as 8th signal type
- [ ] Add configuration examples for common use cases

### 8. Testing

- [ ] Unit tests for `TokenCount` parsing (K/M suffixes, decimals)
- [ ] Unit tests for `ContextClassifier.Classify()`
- [ ] Integration tests for context-based routing decisions
- [ ] E2E tests with different token count ranges

## Reference

- Signal system: `src/semantic-router/pkg/config/config.go` (lines 28-37, 268-293)
- Decision engine: `src/semantic-router/pkg/decision/engine.go` (lines 56-190)
- Classifier: `src/semantic-router/pkg/classification/classifier.go` (lines 596-720)
- Existing signals documentation: `website/docs/overview/signal-driven-decisions.md`
- Metrics: `src/semantic-router/pkg/observability/metrics/metrics.go`

### PR DESCRIPTION
**Pull request by nabobery**
## Summary

This PR implements Issue #1106: Add Context Signal for Token Count-Based Routing.

FIX #1106

### Core Implementation
- **New Signal Type**: Added `context` as the 8th signal type for signal-driven routing
- **TokenCount Type**: Custom type with K/M suffix parsing (e.g., "1K" = 1000, "1.5M" = 1500000)
- **ContextRule Configuration**: Define token count ranges with `min_tokens` and `max_tokens`
- **Parallel Evaluation**: Context signals evaluated alongside other signal types
- **Headers & Observability**: 
    - Added headers `x-vsr-matched-context` and `x-vsr-context-token-count` to requests/responses
    - Added Prometheus histogram `llm_context_token_count`
- **API Response**: Updated `RouterIntentResponse` to include the `context` field in matched signals.

### Dashboard & Frontend
- **Playground**: Added `HeaderDisplay` and `HeaderReveal` support for Context Signal, showing "Signal: Context" and "Token Count" badges.
- **Topology View**: 
    - Updated backend handler to parse and pass context signals to the frontend.
    - Visualizes `context` signals with a new icon (📏) and color scheme in the decision graph.
    - Updated config parser to handle `context_rules` from config.

### Configuration Example
```yaml
signals:
  context_rules:
    - name: "low_token_count"
      min_tokens: "0"
      max_tokens: "1K"
    - name: "high_token_count"
      min_tokens: "4K"
      max_tokens: "128K"

decisions:
  - name: "long_context_handler"
    rules:
      operator: "AND"
      conditions:
        - type: "context"
          name: "high_token_count"
        - type: "domain"
          name: "code"
    modelRefs:
      - model: "claude-3-opus"
```

### Files Changed
| Category | File | Change |
|----------|------|--------|
| **Backend** | `pkg/config/config.go` | Add TokenCount, ContextRule, SignalTypeContext |
| | `pkg/classification/context_classifier.go` | **NEW** - Token counting & classification logic |
| | `pkg/classification/classifier.go` | Integrate ContextClassifier |
| | `pkg/decision/engine.go` | Add `case "context"` to rule evaluation |
| | `pkg/observability/metrics/metrics.go` | Add `llm_context_token_count` histogram |
| | `pkg/headers/headers.go` | Add `x-vsr-matched-context` constants |
| | `pkg/services/classification.go` | Expose context signals in API response structs |
| | `dashboard/backend/handlers/topology.go` | Parse context signals in dashboard backend |
| **Frontend** | `dashboard/frontend/src/components/HeaderDisplay.tsx` | Add context signal UI support |
| | `dashboard/frontend/src/pages/topology/*` | Add context signal to topology visualization |
| **CLI** | `src/vllm-sr/cli/merger.py` | CLI support for context rules |
| **Docs** | `website/docs/tutorials/intelligent-route/context-routing.md` | New tutorial for context routing |

### Testing
- [x] Unit tests for TokenCount parsing (K/M suffixes, decimals, edge cases)
- [x] Unit tests for ContextClassifier.Classify()
- [x] Integration tests for decision engine with context conditions
- [x] Frontend checks (`make dashboard-type-check`)
- [x] Verified in Playground (headers appear) and Topology view (nodes render)

---

- [x] Make sure the code changes pass the [pre-commit](https://github.com/vllm-project/semantic-router/blob/main/CONTRIBUTING.md) checks.
- [x] Sign-off your commit by using `-s` when doing `git commit`
- [x] Try to classify PRs for easy understanding of the type of changes, such as `[Bugfix]`, `[Feat]`, and `[CI]`.

### CODE REVIEW

**Comment by Xunzhuo** [comment]
thanks! can u fix the CI?

**Comment by nabobery** [comment]
Hey @Xunzhuo, will fix the ci and update relevant docs

File: src/semantic-router/pkg/classification/context_classifier.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+// CandleTokenCounter implements TokenCounter using the Candle binding tokenizer
+type CandleTokenCounter struct{}
+
+// CountTokens counts the number of tokens in the given text using the initialized Candle model
+func (c *CandleTokenCounter) CountTokens(text string) (int, error) {
+	// Use a large max length to ensure we count all tokens (up to 1M)
+	// This avoids truncation for most practical context sizes
+	const MaxTokenLimit = 1000000
+	result, err := candle_binding.TokenizeText(text, MaxTokenLimit)
```
Comment (rootfs) [inline src/semantic-router/pkg/classification/context_classifier.go]: tokenization large prompt is an overhead, can you use a quick heuristics like this https://help.openai.com/en/articles/4936856-what-are-tokens-and-how-to-count-them

**Comment by rootfs** [comment]
This is a prompt token filter, it is useful heuristics especially for prefill heavy workload.

File: src/semantic-router/pkg/classification/context_classifier.go (hunk 1 above)
Comment (nabobery) [reply src/semantic-router/pkg/classification/context_classifier.go]: Sure

**Comment by Xunzhuo** [comment]
https://github.com/vllm-project/semantic-router/pull/1105#issuecomment-3771185649
https://github.com/vllm-project/semantic-router/pull/1105#issuecomment-3771187234
can you check these 2 comments as well? @nabobery and rebase the PR : )

**Comment by nabobery** [comment]
> [#1105 (comment)](https://github.com/vllm-project/semantic-router/pull/1105#issuecomment-3771185649) [#1105 (comment)](https://github.com/vllm-project/semantic-router/pull/1105#issuecomment-3771187234) can you check these 2 comments as well? @nabobery and rebase the PR : )

@Xunzhuo  Do you mean the PR description and verifying it on playground?

**Comment by Xunzhuo** [comment]
We need to make sure relevant places have all implemented, in router, playground and topology

**Comment by Xunzhuo** [comment]
You should see the context signal populated when you chat in playground after all the things don

**Comment by nabobery** [comment]
image
@Xunzhuo Tried it on the topology tab of the dashboard using a test query and you could check that the context signals matched

**Comment by Xunzhuo** [comment]
And can you check the playground as well, you need to add the header virtualization like other signal

**Comment by nabobery** [comment]
image
@Xunzhuo Playground using Mock Vllm
